# Snakefile:
# This is the main Snakefile for the project

# Load the environment
configfile: 'environment.yml'

rule all:
    input:
        "results/07_CONTREE/complete",
        "results/07_CONTREE/minOut"
        "results/07_CONTREE/proteinortho"


rule create_environment:
    shell:
        "mamba create -f environment.yml && mamba activate malaria"

rule run_qc_pipeline:
    input:
        raw_data="data/00_raw"
    output:
        qc_folder="results/01_QC"
    shell:
        """
        mkdir -p {output.qc_folder}
        source bin/QC.sh -d {input.raw_data} -o {output.qc_folder}
        """

rule run_gene_prediction:
    input:
        raw_data="data/00_raw"
    output:
        gene_folder_raw="results/02_gene_prediction/raw"
    shell:
        """
        mkdir -p {output.gene_folder_raw}
        source bin/genePred.sh -d {input.raw_data} -o {output.gene_folder_raw} -m 1000
        """

rule clean_genome_assembly:
    input:
        raw_data="data/00_raw/Haemoproteus_tartakovskyi.genome",
        gene_folder="results/02_gene_prediction/raw/genemark.Ht.gtf",
        uniprot_db="data/01_uniprot/aves_taxid.tsv"
    output:
        clean_data="data/02_clean"
    shell:
        """
        mkdir -p {output.clean_data}
        source bin/cleanAves.sh -fa {input.raw_data} -g {input.gene_folder} -db {input.uniprot_db} -gc 27 -o {output.clean_data}
        """

rule rerun_gene_prediction:
    input:
        clean_data="data/02_clean"
    output:
        gene_folder_clean="results/02_gene_prediction/clean"
    shell:
        """
        mkdir -p {output.gene_folder_clean}
        source bin/genePred.sh -d {input.clean_data} -o {output.gene_folder_clean} -m 3000
        """

rule get_orthologous_gene_groups_and_check_busco:
    input:
        clean_data="data/02_clean",
        gene_folder_clean="results/02_gene_prediction/clean",
        busco_db="data/03_busco_downloads/apicomplexa_odb10"
    output:
        busco_folder="results/04_BUSCO",
        ortho_folder="results/05_ORTHO"
    shell:
        """
        mkdir -p {output.busco_folder}
        mkdir -p {output.ortho_folder}
        source bin/getOrtho.sh -i {input.clean_data} -g {input.gene_folder_clean} -d {input.busco_db} -b {output.busco_folder} -o {output.ortho_folder}
        """

rule generate_phylogenetic_tree_from_BUSCO:
    input:
        busco_folder="results/04_BUSCO",
        ortho_folder="results/05_ORTHO"
    output:
        phylo_folder="results/06_PHYLO"
    shell:
        """
        mkdir -p {output.phylo_folder}
        source bin/createPhyloTree.sh -i {input.busco_folder} -p {input.ortho_folder} -o {output.phylo_folder}
        """

rule generate_phylogenetic_tree_from_BUSCO:
    input:
        ortho_folder="results/05_ORTHO"
        ortho_file="Malaria_phylo.proteinortho.tsv"
    output:
        phylo_folder="results/06_PHYLO"
    shell:
        """
        source bin/createPhyloTree.sh -i {input.ortho_folder} -p {input.ortho_file} -o {output.phylo_folder}
        """

rule create_consensus_tree:
    input:
        full_tree="results/06_PHYLO/tree",
        noTG_tree="results/06_PHYLO/noTG_tree"
        proteinortho_tree="results/06_PHYLO/proteinortho_tree"
    output:
        consensus_complete="results/07_CONTREE/complete",
        consensus_minOut="results/07_CONTREE/minOut"
        consensus_proteinortho="results/07_CONTREE/proteinortho"
    shell:
        """
        mkdir -p {output.consensus_complete}
        source bin/consensusTree.sh -i {input.full_tree} -o {output.consensus_complete} -m false
        mkdir -p {output.consensus_proteinortho}
        source bin/consensusTree.sh -i {input.proteinortho_tree} -o {output.consensus_proteinortho} -m false
        mkdir -p {output.consensus_minOut}
        source bin/consensusTree.sh -i {input.noTG_tree} -o {output.consensus_minOut} -m true
        """